File: 백준/최단경로/11657(타임머신).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
const int INF = 987654321;
struct edge {
	int u;
	int v;
	int w;

	edge(int u, int v, int w) {
		this->u = u;
		this->v = v;
		this->w = w;
	}
};

bool bellmanFord(vector<edge>& edgeList, vector<long long>& dist, int N, int M) {
	dist[1] = 0;
	int from, to, weight;
	for (int i = 0; i < N - 1; ++i) {
		for (int j = 0; j < M; ++j) {
			from = edgeList[j].u;
			to = edgeList[j].v;
			weight = edgeList[j].w;

			if (dist[from] == INF) {
				continue;
			}
			dist[to] = min(dist[to], dist[from] + weight);
		}
	}

	bool isCycle = false;
	for (int i = 0; i < M; ++i) {
		from = edgeList[i].u;
		to = edgeList[i].v;
		weight = edgeList[i].w;

		if (dist[from] == INF) {
			continue;
		}
		if (dist[to] > dist[from] + weight) {
			isCycle = true;
			break;
		}
	}
	return isCycle;
}
void input(int& N, int& M, vector<edge>& edgeList) {
	cin >> N >> M;
	int A, B, C;
	for (int i = 0; i < M; ++i) {
		cin >> A >> B >> C;
		edgeList.push_back(edge(A, B, C));
	}
}
int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);

	int N, M;
	vector<edge> edgeList; 
	input(N, M, edgeList);

	vector<long long> dist(N + 1, INF);
	bool isCycle = bellmanFord(edgeList, dist, N, M);
	
	if (isCycle) {
		cout << -1;
	}
	else {
		for (int i = 2; i <= N; ++i) {
			if (dist[i] == INF) {
				cout << -1 << '\n';
			}
			else {
				cout << dist[i] << '\n';
			}
		}
	}
	return 0;
}

File: 백준/최단경로/11657(타임머신).java
import java.io.*;
import java.util.*;

public class Main {

	private static int N;
	private static int M;
	private static Edge[] edges;

	public static void main(String[] args) throws IOException {
		input();

		long[] dists = bellmanFord();

		printAnswer(isCycle(dists), dists);
	}

	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer tokens = new StringTokenizer(br.readLine());
		N = Integer.parseInt(tokens.nextToken());
		M = Integer.parseInt(tokens.nextToken());

		edges = new Edge[M];
		for (int i = 0; i < M; ++i) {
			tokens = new StringTokenizer(br.readLine());
			edges[i] = new Edge(
				Integer.parseInt(tokens.nextToken()),
				Integer.parseInt(tokens.nextToken()),
				Integer.parseInt(tokens.nextToken())
			);
		}
	}

	private static long[] bellmanFord() {
		long[] dists = new long[N + 1];
		Arrays.fill(dists, Long.MAX_VALUE);
		dists[1] = 0;

		for (int i = 0; i < N - 1; ++i) {
			for (int j = 0; j < M; ++j) {
				int from = edges[j].v;
				int to = edges[j].u;
				int weight = edges[j].w;

				if (dists[from] == Long.MAX_VALUE) {
					continue;
				}

				dists[to] = Math.min(dists[to], dists[from] + weight);
			}
		}

		return dists;
	}

	private static boolean isCycle(long[] dists) {
		//음의 cycle 확인
		for (int i = 0; i < M; ++i) {
			int from = edges[i].v;
			int to = edges[i].u;
			int weight = edges[i].w;

			if (dists[from] == Long.MAX_VALUE) {
				continue;
			}

			if (dists[to] > dists[from] + weight) {
				return true;
			}
		}
		return false;
	}

	private static void printAnswer(boolean isCycle, long[] dists) {
		if (isCycle) {
			System.out.println(-1);
		} else {
			for (int i = 2; i < dists.length; ++i) {
				if (dists[i] == Long.MAX_VALUE) {
					dists[i] = -1;
				}
				System.out.println(dists[i]);
			}
		}
	}

	static class Edge {
		int v, u, w;

		Edge(int v, int u, int w) {
			this.v = v;
			this.u = u;
			this.w = w;
		}
	}
}
